src/Functions/overlay.cpp

```diff
@@ -0,0 +1,721 @@
+            if (offset_is_const)
+            {
+                OVERLAY_EXECUTE_CASE(true, true, false)
+            }
+            else
+            {
+                OVERLAY_EXECUTE_CASE(true, false, false)
+            }
+        }
+        else
+        {
+            if (offset_is_const && length_is_const)
+            {
+                OVERLAY_EXECUTE_CASE(false, true, true)
+            }
+            else if (offset_is_const && !length_is_const)
+            {
+                OVERLAY_EXECUTE_CASE(false, true, false)
+            }
+            else if (!offset_is_const && length_is_const)
+            {
+                OVERLAY_EXECUTE_CASE(false, false, true)
+            }
+            else
+            {
+                OVERLAY_EXECUTE_CASE(false, false, false)
+            }
+        }
+#undef OVERLAY_EXECUTE_CASE
+
+        return res_col;
+    }
+
+
+private:
+    /// input offset is 1-based, maybe negative
+    /// output result is 0-based valid offset, within [0, input_size]
+    static size_t getValidOffset(Int64 offset, size_t input_size)
+    {
+        if (offset > 0)
+        {
+            if (static_cast<size_t>(offset) > input_size + 1) [[unlikely]]
+                return input_size;
+            else
+                return offset - 1;
+        }
+        else
+        {
+            if (input_size < -static_cast<size_t>(offset)) [[unlikely]]
+                return 0;
+            else
+                return input_size + offset;
+        }
+    }
+
+    /// get character count of a slice [data, data+bytes)
+    static size_t getSliceSize(const UInt8 * data, size_t bytes)
+    {
+        if constexpr (is_utf8)
+            return UTF8::countCodePoints(data, bytes);
+        else
+            return bytes;
+    }
+
+    template <bool has_three_args, bool offset_is_const, bool length_is_const>
+    void constantConstant(
+        size_t rows,
+        const StringRef & input,
+        const StringRef & replace,
+        const ColumnPtr & column_offset,
+        const ColumnPtr & column_length,
+        Int64 const_offset,
+        Int64 const_length,
+        ColumnString::Chars & res_data,
+        ColumnString::Offsets & res_offsets) const
+    {
+        if (!has_three_args && length_is_const && const_length < 0)
+        {
+            constantConstant<true, offset_is_const, false>(
+                rows, input, replace, column_offset, column_length, const_offset, -1, res_data, res_offsets);
+            return;
+        }
+
+        size_t input_size = getSliceSize(reinterpret_cast<const UInt8 *>(input.data), input.size);
+        size_t valid_offset = 0; // start from 0, not negative
+        if constexpr (offset_is_const)
+            valid_offset = getValidOffset(const_offset, input_size);
+
+        size_t replace_size = getSliceSize(reinterpret_cast<const UInt8 *>(replace.data), replace.size);
+        size_t valid_length = 0; // not negative
+        if constexpr (!has_three_args && length_is_const)
```

Comment:
Re naming `has_three_args` vs. `has_four_args` ... this is one of the places where the latter would be prettier (no negation needed).